src/Bindings/LuaStateParams.h

```diff
@@ -0,0 +1,571 @@
+	{
+		typename std::remove_reference<T>::type dummy;
+		if (!a_LuaState.GetStackValue(a_StackPos, dummy))
+		{
+			return Printf("Mismatch, expected %s, got %s",
+				GetTypeDescription<std::remove_reference<T>::type>().c_str(),
+				a_LuaState.GetTypeText(a_StackPos).c_str()
+			);
+		}
+		return AString();
+	}
+
+
+	/** Returns an error message if the value on the specified index on the Lua stack is of the wrong (template) type.
+	This is the specialization for "cNonNil"-decorated values. */
+	template <typename T>
+	static AString CheckValueType(cLuaState & a_LuaState, int a_StackPos, const cNonNil<T> & a_Dest)
+	{
+		if (lua_isnil(a_LuaState, a_StackPos))
+		{
+			return Printf("Expected a non-nil instance of %s, got a nil",
+				GetTypeDescription<std::remove_reference<T>::type>().c_str(),
+				a_LuaState.GetTypeText(a_StackPos).c_str()
+			);
+		}
+		return CheckValueType(a_LuaState, a_StackPos, a_Dest.m_Dest);
+	}
+
+
+	/** Returns an error message if the value on the specified index on the Lua stack is of the wrong (template) type.
+	This is the specialization for "cSelf"-decorated values. */
+	template <typename T>
+	static AString CheckValueType(cLuaState & a_LuaState, int a_StackPos, const cSelf<T> & a_Dest)
+	{
+		if (lua_isnil(a_LuaState, a_StackPos))
+		{
+			return Printf("Expected an instance of %s, got a %s. Did you use the right calling convention?",
+				GetTypeDescription<std::remove_reference<T>::type>().c_str(),
+				a_LuaState.GetTypeText(a_StackPos).c_str()
+			);
+		}
+		return CheckValueType(a_LuaState, a_StackPos, a_Dest.m_Self);
+	}
+
+
+	/** Returns an error message if the value on the specified index on the Lua stack is of the wrong (template) type.
+	This is the specialization for "cStaticSelf"-decorated values. */
+	template <typename T>
+	static AString CheckValueType(cLuaState & a_LuaState, int a_StackPos, const cStaticSelf<T> & a_Dest)
+	{
+		tolua_Error err;
+		auto type = GetTypeDescription<std::remove_pointer<typename std::remove_reference<T>::type>::type>();
+		if (lua_isnil(a_LuaState, a_StackPos))
+		{
+			return Printf("Expected the class %s, got a nil", type.c_str());
+		}
+		if (tolua_isusertype(a_LuaState, a_StackPos, type.c_str(), 0, &err))
+		{
+			return Printf("Expected the class %s, got a %s. This function is static, remember to use the right calling convention.",
+				type.c_str(),
+				a_LuaState.GetTypeText(a_StackPos).c_str()
+			);
+		}
+		return AString();
+	}
+
+
+
+	/** Returns a string describing the function overload composed of the specified (template) types. */
+	template <typename T1, typename T2, typename... T3>
+	static AString BuildSingleOverloadDescription()
+	{
+		return GetTypeDescription<T1>() + ", " + BuildSingleOverloadDescription<T2, T3...>();
+	}
+
+
+	/** Terminator for the above function. */
+	template <typename T>
+	static AString BuildSingleOverloadDescription()
+	{
+		return GetTypeDescription<T>();
+	}
```

Comment:
Again, an empty case is cleaner imo.

Reply:
I don't think I can do that, because the recursion adds a `, ` separator in between the levels, so the last level still needs to return something valid.

Reply:
Ah you're right, I missed that.